### ISSUE TITLE
Ability to do more complex route matching

### ISSUE THREAD

**Issue by mkieblesz** [tokio-rs/axum#109]
### Motivation

- full path match
- multiple groups in a single path segment

### Proposal

There could be a lot of ways of doing this, but the idea is basically the same - allow to pass custom regex to each capture group in route string. Imo best would be to allow to specify name of the the passed group (lookup custom converter in flask). These groups would be predefined in some sort of CustomConverterMap<string, Regex> and somehow passed to the route function (wip).

```rust
# how it is now
route("/greet/:id", handler)

# how it could be
route("/greet/<int:id>", handler)
```

How others are doing it: 
- flask way: https://werkzeug.palletsprojects.com/en/2.0.x/routing/#custom-converters
- django way: https://docs.djangoproject.com/en/3.2/ref/urls/#re-path
- rails way: https://guides.rubyonrails.org/routing.html#segment-constraints

### Alternatives

Probably the only alternative is changing the spec for url patterns of the application.

**Comment by davidpdrsn** [comment]
Note to self: Make sure we don't add something that might conflict with https://github.com/tokio-rs/axum/issues/50

**Comment by mkatychev** [comment]
To follow up on https://github.com/tokio-rs/axum/pull/122#event-5117809154 here. Would it be reasonable to accommodate enums for the route path? 

I think regex is a very solid foundation but it would be great to allow matching on request method + uri. I realize looping through route aliases is probably not the most efficient way of doing grouping so it would be good to have some functional control of path + method similar to how `Request<Body>`:
```rust
match (req.method(), req.uri().path()) {
    (Method::GET, "/" | "/home") => Some("/" get(home_handler)),
    (Method::POST, "/" | "/home") => Some("/" post(home_handler)),
    _ => (None),
}
```

This can also help avoid precedent logic: https://github.com/tokio-rs/axum/issues/62#issuecomment-890380738.

I think the current pattern of `.route(path, method(handler))` too strictly bisects the routing logic. Something silimar to how the `Handler` trait allows closures to be passed in would go a long way into allowing simple yet flexible routing:
```rust
let app = route("/", get(|| async { Ok::<_, Infallible>(StatusCode::OK) }));
```

**Comment by davidpdrsn** [comment]
Here are my thoughts:

axum's current routing consisting of fixed paths (`route("/foo/bar", ...)`) or named parameters (`route("/users/:id", ...)`) is honestly all I've ever needed, so thats why axum shipped with that. However I am open to adding more flexible routing and agree that it has great utility.

axum's current router is quite naive. It does a linear search through all your routes until it finds one that matches. I went with this design because its easy to understand and implement, fits well with axum's general design of wrapping services in other services, and probably is fast enough for most users.

Its important to me that whatever additional routing features we add don't prevent us from optimizing the router in the future. Thats why I'm not fond of supporting general regexes for routes. Regexes also have maintenance issues -- they're often hard to understand 6 months later.

My plan is to do some research and look at the routing features provides by other Rust frameworks and then make a decision on what makes sense for axum. As mentioned above, whatever feature set we pick should be compatible with generating OpenAPI docs.

Tbh I dunno when I'll get to that. I currently think there are more important problems to solve but I encourage people to keep posting here with features they would like.

Small aside: I think wildcard routes like `/foo/*/bar` are great and I think they provide a good solution for problems like https://github.com/tokio-rs/axum/discussions/99. So I'd like to support that as well.

**Comment by kevlarr** [comment]
> axum's current routing consisting of fixed paths (route("/foo/bar", ...)) or named parameters (route("/users/:id", ...)) is honestly all I've ever needed, so thats why axum shipped with that. However I am open to adding more flexible routing and agree that it has great utility.
>
>axum's current router is quite naive. It does a linear search through all your routes until it finds one that matches. I went with this design because its easy to understand and implement, fits well with axum's general design of wrapping services in other services, and probably is fast enough for most users.
>
> Its important to me that whatever additional routing features we add don't prevent us from optimizing the router in the future. Thats why I'm not fond of supporting general regexes for routes. Regexes also have maintenance issues -- they're often hard to understand 6 months later.

I'm assuming those points mean we won't eventually see an option to extract multiple named parameters from the same path segment, eg. `"/some/path/{one}:{two}-{three}"` if we want the framework to validate some format and extract into separate variables, rather than doing in-code?

**Comment by davidpdrsn** [comment]
@kevlarr Its unlikely. That seems like a pretty niche feature that is well served by capturing stuff into a single param and parsing it in the handler. With `extract::Path` you could even write a deserializer to have it done transparently behind the scenes.

**Comment by kevlarr** [comment]
Yeah, definitely a pretty niche use case - I've more typically seen features like that advertised when there are explicit patterns (eg. extensions) that shouldn't be captured (eg. `"/some/{resource}.json"`), but I can't recall actually using a router like that so am not really *that* convinced that there are any advantages to it, especially given:

> With extract::Path you could even write a deserializer to have it done transparently behind the scenes.

**Comment by kevlarr** [comment]
Aside, comments like...

> However I am open to adding more flexible routing and agree that it has great utility.

> My plan is to do some research and look at the routing features provides by other Rust frameworks and then make a decision on what makes sense for axum

> ... but I encourage people to keep posting here with features they would like.

... are commendable and really nice/encouraging to see from a project owner. Thanks for being thoughtful/reflective/open like that.

**Comment by davidpdrsn** [comment]
Once 0.2 is shipped (probably in a week or two) I'll start thinking more concretely about routing. Just have to make sure it's compatible with openapi.

Will of course post here and gather feedback before coding anything.

**Comment by scott-ling** [comment]
agree that a * wildcard is needed, never been convinced that a URI path is complex enough for a regex (plus regex is slow)

but a catch-all for all routes that fan out from a base route for example /*  and /my/resources/* for example are quite useful across a varied set of use cases.
